ast27/Python/ast.c

```diff
@@ -848,15 +853,32 @@ ast_for_arguments(struct compiling *c, const node *n)
                 vararg = NEW_IDENTIFIER(CHILD(n, i+1));
                 if (!vararg)
                     return NULL;
-                i += 3;
+                i += 2 + (TYPE(CHILD(n, i + 2)) == COMMA);
                 break;
             case DOUBLESTAR:
                 if (!forbidden_check(c, CHILD(n, i+1), STR(CHILD(n, i+1))))
                     return NULL;
                 kwarg = NEW_IDENTIFIER(CHILD(n, i+1));
                 if (!kwarg)
                     return NULL;
-                i += 3;
+                i += 2 + (TYPE(CHILD(n, i + 2)) == COMMA);
+                break;
+            case TYPE_COMMENT:
+                assert(l < k + !!vararg + !!kwarg);
+
+                if (!type_comments) {
+                    /* lazily allocate the type_comments seq for perf reasons */
+                    type_comments = asdl_seq_new(n_all_args, c->c_arena);
+                    if (!type_comments)
+                        return NULL;
+                }
+
+                while (l < k + !!vararg + !!kwarg - 1) {
+                    asdl_seq_SET(type_comments, l++, NULL);
+                }
+
+                asdl_seq_SET(type_comments, l++, NEW_TYPE_COMMENT(ch));
+                i += 1;
```

Comment:
All this logic looks sensible.

I'd like to have a comment somewhere describing the intended semantics of this `type_comment` sequence -- something like that it's either `NULL`, or has one entry for each parameter including one for varargs and one for kwargs if applicable, any of which may be `NULL` for an absent type comment.